### PR TITLE
Fix WhenAllOrAnyFailed to not eat OperationCanceledExceptions

### LINKED ISSUE #58081: TaskTimeoutExtensions.WhenAllOrAnyFailed eats OperationCanceledExceptions from awaited tasks
If one of the tasks passed to WhenAllOrAnyFailed throws OperationCanceledException, the original exception is eaten and replaced with a new cancellation exception that comes directly from WhenAllOrAnyFailed itself.

Issue seems to be caused by this: https://github.com/dotnet/runtime/blob/main/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs#L110

And this logic looks suspect too: https://github.com/dotnet/runtime/blob/main/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs#L82

Seems like we should not treat a cancellation exception any differently here than any other exception.

@stephentoub Thoughts?

### PR DESCRIPTION
**Pull request by geoffkizer**
Fixes https://github.com/dotnet/runtime/issues/58081

### CODE REVIEW

File: src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs [hunk 1]
```diff
@@ -95,28 +94,37 @@ public static async Task WhenAllOrAnyFailed(this Task[] tasks)
         private static Task WhenAllOrAnyFailedCore(this Task[] tasks)
         {
             int remaining = tasks.Length;
-            var tcs = new TaskCompletionSource<bool>();
+            var tcs = new TaskCompletionSource();
```
Comment (stephentoub) [inline src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:97]: Does this file not get used in any tests that target .NET Framework, e.g. with WinHttpHandler?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs (hunk 1 above)
Comment (geoffkizer) [reply src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:97]: It probably does.

**Review by stephentoub — APPROVED** [review]

**Comment by geoffkizer** [comment]
Failure looks unrelated.

**Comment by ChrML** [comment]
Shouldn't there be a test to verify this in this PR?

**Comment by stephentoub** [comment]
This function is only used by tests.

**Pull request merged**